r3bgen/R3BParticleGenerator.h

```diff
@@ -0,0 +1,103 @@
+/******************************************************************************
+ *   Copyright (C) 2019 GSI Helmholtzzentrum für Schwerionenforschung GmbH    *
+ *   Copyright (C) 2019 Members of R3B Collaboration                          *
+ *                                                                            *
+ *             This software is distributed under the terms of the            *
+ *                 GNU General Public Licence (GPL) version 3,                *
+ *                    copied verbatim in the file "LICENSE".                  *
+ *                                                                            *
+ * In applying this license GSI does not waive the privileges and immunities  *
+ * granted to it by virtue of its status as an Intergovernmental Organization *
+ * or submit itself to any jurisdiction.                                      *
+ ******************************************************************************/
+
+#ifndef R3BGENERATOR_H
+#define R3BGENERATOR_H
+
+#include "FairGenerator.h"
+#include "R3BDistribution.h"
+#include "R3BDistribution1D.h"
+#include "TRandom3.h"
+
+#include <iostream>
+
+class FairPrimaryGenerator;
+
+// C++17 will make the following lines much better, but for now we have to stick with this...
+// maybe not very good to read, but its at least constexpr
+constexpr int ChargeMass2PDGCode(int z, uint a)
+{
+    return (z > 0 ? (1000000000 + z * 10000 + a * 10) : -(1000000000 - z * 10000 + a * 10));
+}
+constexpr int PDGCode2Mass(int pdg)
+{
+    return (pdg > 1000000000 ? ((pdg / 10) % 1000) : (pdg < 1000000000 ? ((-pdg / 10) % 1000) : 0));
+}
+constexpr int PDGCode2Charge(int pdg)
+{
+    return (pdg > 1000000000 ? ((pdg / 10000) % 1000) : (pdg < 1000000000 ? -((-pdg / 10000) % 1000) : 0));
+}
+constexpr std::array<int, 2> PDGCode2ChargeMass(int pdg) { return { PDGCode2Charge(pdg), PDGCode2Mass(pdg) }; }
+
+class R3BParticleGenerator : public FairGenerator
+{
+
+  public:
+    R3BParticleGenerator();
+
+    R3BParticleGenerator(Int_t pdgCode, UInt_t seed = 0);
+    R3BParticleGenerator(const Char_t* ionName, unsigned int seed = 0);
+
+    void SetProton();
+    void SetNeutron();
+    void SetIon(const Char_t* ionName);
+    void SetParticle(Int_t pdgCode);
+    void SetParticle(Int_t charge, UInt_t mass)
+    {
+        if (mass == 1 && charge == 1)
+            SetProton();
+        else if (mass == 1 && charge == 0)
+            SetNeutron();
+        else
+            SetParticle(ChargeMass2PDGCode(charge, mass));
+    }
+
+    void SetVertexDistribution(R3BDistribution<3> vertexDistribution_x_y_z_cm)
+    {
+        fVertex_cm = vertexDistribution_x_y_z_cm;
+    }
+    R3BDistribution<3>& GetVertexDistribution() { return fVertex_cm; }
+
+    void SetEnergyDistribution(const R3BDistribution<1>& energyDistribution_AMeV)
+    {
+        fEnergy_AMeV = energyDistribution_AMeV;
+    }
+    R3BDistribution<1>& GetEnergyDistribution() { return fEnergy_AMeV; }
```

Comment:
Implicitly, the input is handled as AMeV for Ions and as MeV for particles. This does match the common usage and is the right behavior. Maybe add something to make it more explicit? Or a comment? idk.

Reply:
Yeah I chose actually not change R3BIonGenerator, becasue it produces several ions, each from a different position, and not all from the same like I would have assumed.